app/src/main/java/org/onionshare/android/tor/TorManager.kt

```diff
@@ -27,113 +31,149 @@ import java.io.FileInputStream
-import kotlin.coroutines.suspendCoroutine
+import kotlin.math.roundToInt
 
 private val LOG = getLogger(TorManager::class.java)
 private val EVENTS = listOf(
     EVENT_CIRCUIT_STATUS, // this one is needed for TorService to function
-    EVENT_OR_CONN_STATUS, // can be removed before release
     EVENT_HS_DESC,
-    EVENT_NEW_DESC, // can be removed before release
+    EVENT_STATUS_CLIENT,
     EVENT_WARN_MSG,
     EVENT_ERR_MSG,
 )
+private val BOOTSTRAP_REGEX = Regex("^NOTICE BOOTSTRAP PROGRESS=([0-9]{1,3}) .*$")
 
 @Singleton
 class TorManager @Inject constructor(
     private val app: Application,
 ) {
-    private var broadcastReceiver: BroadcastReceiver? = null
+    private val _state = MutableStateFlow<TorState>(TorState.Stopped)
+    internal val state: StateFlow<TorState> = _state
+    private var broadcastReceiver = object : BroadcastReceiver() {
+        /**
+         * Attention: This gets executes on UI Thread
+         */
+        override fun onReceive(context: Context, i: Intent) {
+            when (i.getStringExtra(EXTRA_STATUS)) {
+                TorService.STATUS_STARTING -> LOG.debug("TorService: Starting...")
+                TorService.STATUS_ON -> {
+                    LOG.debug("TorService: Started")
+                    startLatch?.countDown() ?: error("startLatch was null when Tor started.")
+                }
+                // FIXME When we stop unplanned, we need to inform the ShareManager
+                //  that we stopped, so it can clear its state up, stopping webserver, etc.
+                TorService.STATUS_STOPPING -> LOG.debug("TorService: Stopping...")
+                TorService.STATUS_OFF -> LOG.debug("TorService: Stopped")
+            }
+        }
+    }
+
+    @Volatile
+    private var broadcastReceiverRegistered = false
+    private var startLatch: CountDownLatch? = null
+
+    @Volatile
+    private var controlConnection: TorControlConnection? = null
+    private val onionListener = RawEventListener { keyword, data ->
+        // TODO consider removing the logging below before release
+        LOG.debug("$keyword: $data")
+        // bootstrapping gets 70% of our progress
+        if (keyword == EVENT_STATUS_CLIENT) {
+            val matchResult = BOOTSTRAP_REGEX.matchEntire(data)
+            val percent = matchResult?.groupValues?.get(1)?.toIntOrNull()
+            if (percent != null) {
+                val progress = (percent * 0.7).roundToInt()
+                val newState = (state.value as? TorState.Starting)?.copy(progress = progress)
+                    ?: TorState.Starting(progress)
+                _state.value = newState
+            }
+            return@RawEventListener
+        }
+        val onion = (state.value as? TorState.Starting)?.onion
+        // descriptor upload counts as 90%
+        if (onion != null && keyword == EVENT_HS_DESC && data.startsWith("UPLOAD $onion")) {
+            _state.value = TorState.Starting(90, onion)
```

Comment:
Good catch! I *think* then `onion` would be `null`, because the state wasn't `TorState.Starting`, but this is rather implicit, so I'll add an extra check to make it more explicit.